File: src/lib/tauri/googleSheets.ts
import { invoke } from "@tauri-apps/api/core";

import { CMD } from "./commands";
import type { ColumnMeta, QueryResult } from "./types";

export type GoogleSheetInfo = {
  title: string;
  row_count: number;
  column_count: number;
};

export type GoogleSheetsMetadata = {
  spreadsheet_title: string;
  sheets: GoogleSheetInfo[];
};

export type GoogleSheetOverview = {
  columns: ColumnMeta[];
  row_count: number;
};

export function googleSheetsListSheets(
  connectionId: string
): Promise<GoogleSheetsMetadata> {
  return invoke<GoogleSheetsMetadata>(CMD.googleSheetsListSheets, {
    connectionId,
  });
}

export function googleSheetsSheetOverview(args: {
  connectionId: string;
  sheet: string;
}): Promise<GoogleSheetOverview> {
  return invoke<GoogleSheetOverview>(CMD.googleSheetsSheetOverview, args);
}

export async function googleSheetsFetchRows(args: {
  connectionId: string;
  sheet: string;
  limit: number;
  offset: number;
}): Promise<QueryResult> {
  const result = await invoke<{
    columns: QueryResult["columns"];
    rows: QueryResult["rows"];
    row_count: number;
  }>(CMD.googleSheetsFetchRows, args);
  return {
    columns: result.columns ?? [],
    rows: result.rows ?? [],
    rowCount: Number(result.row_count ?? 0),
  };
}

File: src-tauri/src/commands/google_sheets.rs
use tauri::State;
use uuid::Uuid;

use crate::engines::google_sheets::api::{
    fetch_metadata, fetch_overview, fetch_rows, GoogleSheetOverview, GoogleSheetRows,
    GoogleSheetsMetadata,
};
use crate::engines::EngineConnection;
use crate::state::AppState;

fn google_sheets_connection(
    state: &AppState,
    connection_id: Uuid,
) -> Result<crate::engines::google_sheets::connection::GoogleSheetsConn, String> {
    let connection = state
        .connections
        .get(&connection_id)
        .ok_or("CONNECTION_NOT_FOUND")?;
    match connection.value() {
        EngineConnection::GoogleSheets(connection) => Ok(connection.clone()),
        _ => Err("ENGINE_NOT_SUPPORTED".into()),
    }
}

#[tauri::command]
pub async fn google_sheets_list_sheets(
    state: State<'_, AppState>,
    connection_id: Uuid,
) -> Result<GoogleSheetsMetadata, String> {
    let connection = google_sheets_connection(&state, connection_id)?;
    fetch_metadata(
        &connection.http,
        &connection.spreadsheet_id,
        &connection.credential,
    )
    .await
}

#[tauri::command]
pub async fn google_sheets_sheet_overview(
    state: State<'_, AppState>,
    connection_id: Uuid,
    sheet: String,
) -> Result<GoogleSheetOverview, String> {
    let connection = google_sheets_connection(&state, connection_id)?;
    let metadata = fetch_metadata(
        &connection.http,
        &connection.spreadsheet_id,
        &connection.credential,
    )
    .await?;
    let sheet_info = metadata
        .sheets
        .iter()
        .find(|item| item.title == sheet)
        .ok_or("GOOGLE_SHEETS_WORKSHEET_NOT_FOUND")?;
    fetch_overview(
        &connection.http,
        &connection.spreadsheet_id,
        &connection.credential,
        &sheet,
        sheet_info.row_count,
        sheet_info.column_count,
    )
    .await
}

#[tauri::command]
pub async fn google_sheets_fetch_rows(
    state: State<'_, AppState>,
    connection_id: Uuid,
    sheet: String,
    limit: usize,
    offset: usize,
) -> Result<GoogleSheetRows, String> {
    let connection = google_sheets_connection(&state, connection_id)?;
    fetch_rows(
        &connection.http,
        &connection.spreadsheet_id,
        &connection.credential,
        &sheet,
        limit.clamp(1, 1_000),
        offset,
    )
    .await
}
